web/modules/slavnews_weather/src/Plugin/Block/WeatherBlockFolder/WeatherBlock.php

```diff
@@ -23,39 +68,102 @@ class WeatherBlock extends BlockBase {
    * Building function.
    */
   public function build() {
-    // (receive weather api)
-    $method = 'GET';
-    $request = new Request();
-    // (receive client ip)
-    $ip = $request->getClientIp();
-    $ip_query = 'auto:' . $ip;
-    // (receive weather api)
+    $requestedWeatherData = $this->getWeather();
+    return $requestedWeatherData;
+  }
+
+  /**
+   * Receive all weather data to display.
+   */
+  public function getWeather() {
+    // Setting cid for cache that is saved.
+    $cacheId = "slavwnews_weather";
+    // Function that request's user geological location by ip.
+    $location = $this->getIplocation();
+
+    // If we have data in cache we do following.
+    if ($cache = $this->cacheBackend->get($cacheId)) {
+      // Get data from cache.
+      $data = $cache->data;
+      // Return data.
+      return [
+        '#theme' => 'slavnews__weather__block',
+        '#image' => $data['current']['condition']['icon'],
+        '#temp_c' => $data['current']['temp_c'],
+        '#city' => $data['location']['name'],
+      ];
+    }
```

Comment:
What if the first user gets the weather from Kovel and another user from Lutsk opens the site within 1min later?